### ISSUE TITLE
cfreactor breaks twisted.trial

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2556]
|<img alt="bromine's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bromine reported|
|-|-|
|Trac ID|trac#2556|
|Type|defect|
|Created|2007-04-09 17:42:55Z|

I have run into a number of problems trying to use cfreactor with twisted.trial. 

The biggest problem is that, out of the box, using cfreactor causes twisted.trial to hang. To reproduce this problem, you can put [[test.py](None/test.py) test.py] and [[testcase.py](None/testcase.py) testcase.py] into a folder and then run 

```
python test.py
```

from that folder. Python will hang in test.py, which just calls twisted.trial with --reactor cf testcase.py.

This problem is corrected by the [[cfreactor.py.diff](None/cfreactor.py.diff) attached diff]. (The issue is that self.didStartRunLoop has to be set before calling runLoop.run() in order for the correct value of self.didStartRunLoop to be available inside the run loop.)

However, even fixing that still leaves two problems with running trial:

```
twisted/internet/cfreactor.py:294: exceptions.DeprecationWarning: reactor.crash cannot be used inside unit tests. By Twisted 2.7, using crash will fail the test and may crash or hang the test run.

/Software/twisted/trial/reporter.py:219: twisted.trial.reporter.BrokenTestCaseWarning: REACTOR UNCLEAN! traceback(s) follow: 
Traceback (most recent call last):
  File "twisted/trial/unittest.py", line 652, in _classCleanUp
    util._Janitor().postClassCleanup()
  File "twisted/trial/util.py", line 68, in postClassCleanup
    'cleanPending', 'cleanThreads')
  File "twisted/trial/util.py", line 72, in _dispatch
    getattr(self, "do_%s" % attr)()
  File "twisted/trial/util.py", line 126, in do_cleanReactor
    raise DirtyReactorError(' '.join(s))
twisted.trial.util.DirtyReactorError: THIS WILL BECOME AN ERROR SOON! reactor left in unclean state, the following Selectables were left over:  <twisted.internet.posixbase._UnixWaker instance at 0x2a44af8>
```

Both of these are pointing at a mismatch between the semantics of reactors expected by trial and those actually implemented by cfreactor, but I don't know enough about either to know what the right way to fix this is.

Attachments:

* [cfreactor.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cdc/cdc4a83ae484cf1e0aec7b86c9dc6660a6eeea83/96a91984ae2311a1aacc252fe38bf2b34805a419.diff) (531 bytes) - added by bromine on 2007-04-09 17:43:12Z - 
* [test.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cdc/cdc4a83ae484cf1e0aec7b86c9dc6660a6eeea83/a36e355ddb203d7d4133221f339dc406cb9f480f.py) (189 bytes) - added by bromine on 2007-04-09 17:43:36Z - 
* [testcase.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cdc/cdc4a83ae484cf1e0aec7b86c9dc6660a6eeea83/c6fa543491d4da1185be2699b43835c0d5ed8473.py) (274 bytes) - added by bromine on 2007-04-09 17:43:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2556 2556
type__defect defect
reporter__bromine bromine
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1176140575000000 1176140575000000
changetime__1289001324000000 1289001324000000
version__None None
owner__bromine bromine

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Hmm cfreactor.  I wonder if this is a problem that was introduced after it was initially written, or if it has always worked this way.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to bromine|
|-|-|

cfreactor could really use a maintainer.  Any chance you might be willing to take on that role?  Starting with a patch for this issue as per TwistedDevelopment, of course? :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set status to `closed`|
|-|-|

(In [30231]) Fix cfreactor.  Remove cfsupport.

Author: glyph

Reviewer: exarkun

Fixes: [#1833](https://github.com/twisted/twisted/issues/1833)

Fixes: [#2556](https://github.com/twisted/twisted/issues/2556)

Fixes: [#4652](https://github.com/twisted/twisted/issues/4652)

This change significantly improves cfreactor.  It is a rewrite from our broken,
unmaintained 'cfsupport' pyrex/C module (which has been removed) to use
PyObjC's CoreFoundation and CFNetwork support.  This fixes several very serious
bugs, including many segfaults.

This also changes the testing status of cfreactor.  Previously it received no
coverage at all from the buildbots.  While this change does not add any unique
test code, it adds the CFReactor to all tests which use 'reactorbuilder'.
These tests would previously fail (several with segfaults) and they now all
pass.  The result of this is that most of the direct reactor tests are invoked
on our existing OS X buildbot.  It's still not 100% supported, since there is
still no buildbot running under the cfreactor via 'trial -r cf'.  A few tests
still fail intermittently in that configuration, but I believe they are all
bugs in the tests, rather than in the reactor.

This change also allows Twisted to be built with the 'clang' compiler from the
LLVM project.  Since 'cfsupport' was the only code that had problems with that
compiler, its removal fixes the issue.

(This change also includes several unrelated flymake issues.)
